server/src/main/java/org/opensearch/common/blobstore/EncryptedBlobContainer.java

```diff
@@ -0,0 +1,194 @@
+        executeWrite(
+            inputStream,
+            blobSize,
+            (encryptedStream, encryptedLength) -> blobContainer.writeBlobAtomic(
+                blobName,
+                encryptedStream,
+                encryptedLength,
+                failIfAlreadyExists
+            )
+        );
+    }
+
+    @Override
+    public DeleteResult delete() throws IOException {
+        return blobContainer.delete();
+    }
+
+    @Override
+    public void deleteBlobsIgnoringIfNotExists(List<String> blobNames) throws IOException {
+        blobContainer.deleteBlobsIgnoringIfNotExists(blobNames);
+    }
+
+    @Override
+    public Map<String, BlobMetadata> listBlobs() throws IOException {
+        Map<String, BlobMetadata> blobMetadataMap = blobContainer.listBlobs();
+        return convertToEncryptedMetadataMap(blobMetadataMap);
+    }
+
+    @Override
+    public Map<String, BlobContainer> children() throws IOException {
+        Map<String, BlobContainer> children = blobContainer.children();
+        if (children != null) {
+            return children.entrySet()
+                .stream()
+                .collect(Collectors.toMap(Map.Entry::getKey, entry -> new EncryptedBlobContainer<>(entry.getValue(), cryptoHandler)));
+        } else {
+            return null;
+        }
+    }
+
+    @Override
+    public Map<String, BlobMetadata> listBlobsByPrefix(String blobNamePrefix) throws IOException {
+        Map<String, BlobMetadata> blobMetadataMap = blobContainer.listBlobsByPrefix(blobNamePrefix);
+        return convertToEncryptedMetadataMap(blobMetadataMap);
+    }
+
+    private Map<String, BlobMetadata> convertToEncryptedMetadataMap(Map<String, BlobMetadata> blobMetadataMap) {
+        if (blobMetadataMap == null) {
+            return null;
+        }
+
+        return blobMetadataMap.entrySet()
+            .stream()
+            .collect(
+                Collectors.toMap(
+                    Map.Entry::getKey,
+                    entry -> new EncryptedBlobMetadata(entry.getValue(), cryptoHandler, getEncryptedHeaderContentSupplier(entry.getKey()))
+                )
+            );
+
+    }
+
+    @Override
+    public void listBlobsByPrefixInSortedOrder(
+        String blobNamePrefix,
+        int limit,
+        BlobNameSortOrder blobNameSortOrder,
+        ActionListener<List<BlobMetadata>> listener
+    ) {
+        ActionListener<List<BlobMetadata>> encryptedMetadataListener = ActionListener.delegateFailure(
+            listener,
+            (delegatedListener, metadataList) -> {
+                if (metadataList != null) {
+                    List<BlobMetadata> encryptedMetadata = metadataList.stream()
+                        .map(
+                            blobMetadata -> new EncryptedBlobMetadata(
+                                blobMetadata,
+                                cryptoHandler,
+                                getEncryptedHeaderContentSupplier(blobMetadata.name())
```

Comment:
Didn't understand exactly what you mean here. If you are saying that if there's a way to avoid reads of a file header to calculate length then answer is no. Length of decrypted content of a file is dependent on encryption metadata and encryption metadata is kept in file header. This metadata is constant today but can vary on a number of factors in future. Few examples are upgrades of encryption SDK, support for data keys of different lengths, different frame sizes and support for different algorithms. If we don't have right metadata then we will lose the content.

Reply:
Got it. Then in that case we would need to introduce another `list` api in repo which returns the list of names instead of metadata .